api/routes/DoorCode.js

```diff
@@ -0,0 +1,109 @@
+const express = require('express');
+const router = express.Router();
+const DoorCode = require('../models/DoorCode');
+const settings = require('../util/settings');
+const logger = require(`${settings.util}/logger`);
+const {
+  checkIfTokenSent,
+  checkIfTokenValid,
+} = require('../util/token-functions');
+const {
+  OK,
+  BAD_REQUEST,
+  UNAUTHORIZED,
+  FORBIDDEN,
+  NOT_FOUND,
+} = require('../constants').STATUS_CODES;
+const addErrorLog = require('../util/logging-helpers');
+
+router.get('/getAvailableCodes', (req, res) => {
+  DoorCode.find({ usersAssigned: { $lt: 2 } }).then((doorcodes) =>
+    res.status(OK).send(doorcodes)
+  );
+});
+
+router.post('/addCode', (req, res) => {
+  if (!checkIfTokenSent(req)) {
+    return res.sendStatus(FORBIDDEN);
+  } else if (!checkIfTokenValid(req)) {
+    return res.sendStatus(UNAUTHORIZED);
+  }
+
+  const newCode = new DoorCode({
+    dateLastUsed: req.body.dateLastUsed,
+    doorCode: req.body.doorCode,
+    doorCodeValidUntil: req.body.doorCodeValidUntil,
+    usersAssigned: req.body.usersAssigned,
+  });
+
+  DoorCode.create(newCode, (error, post) => {
+    if (error) {
+      logger.log(`Event /addCode error: ${error}`);
+      return res.sendStatus(BAD_REQUEST);
+    }
+    return res.status(OK).send(post);
+  });
+});
+
+router.post('/editCode', (req, res) => {
+  if (!checkIfTokenSent(req)) {
+    return res.sendStatus(FORBIDDEN);
+  } else if (!checkIfTokenValid(req)) {
+    return res.sendStatus(UNAUTHORIZED);
+  }
+
+  const {
+    dateLastUsed,
+    doorCode,
+    doorCodeValidUntil,
+    usersAssigned,
+  } = req.body;
+  DoorCode.findOne({ doorCode: req.body.oldDoorCode })
+    .then((code) => {
+      code.dateLastUsed = dateLastUsed || code.dateLastUsed;
+      code.doorCode = doorCode || code.doorCode;
+      code.doorCodeValidUntil = doorCodeValidUntil || code.doorCodeValidUntil;
+      code.usersAssigned = usersAssigned || code.usersAssigned;
+      code
+        .save()
+        .then((ret) => {
+          res.status(OK).json({ ret, code: 'door code updated successfully' });
+        })
+        .catch((error) => {
+          res.status(BAD_REQUEST).send({
+            error,
+            message: 'door code was not updated',
+          });
+        });
+    })
+    .catch((error) => {
+      res.status(NOT_FOUND).send({ error, message: 'door code not found' });
+    });
+});
+
+router.post('/removeCode', (req, res) => {
+  if (!checkIfTokenSent(req)) {
+    return res.sendStatus(FORBIDDEN);
+  } else if (!checkIfTokenValid(req)) {
+    return res.sendStatus(UNAUTHORIZED);
+  }
+
+  DoorCode.deleteOne({ doorCode: req.body.doorCode }, (error, form) => {
```

Comment:
i would instead search by `_id` here, imagine the situation where we have a duplicate code, one with assignees and one without, deleting one may remove members' door codes unintentionally
```suggestion
  DoorCode.findOne({ doorCode: req.body._id })
```

for how to remove with `_id` instead check out this function in [`api/routes/Event.js`](https://github.com/SCE-Development/Core-v4/blob/ea3b2c59b57957225894fea89b1db2d24fbd10c6/api/routes/Event.js#L103) and the associated test case in [`test/Event.js`](https://github.com/SCE-Development/Core-v4/blob/ea3b2c59b57957225894fea89b1db2d24fbd10c6/test/Event.js#L201)